### PR TITLE
ddl: add check when build enum, set column

### PR DESCRIPTION
**Pull request by crazycs520**
### What problem does this PR solve? 
Execute below sql should return error: `Column 'a' has duplicated value xx in ENUM"`.
But currently TiDB will execute successfully.
```sql
create table t1 (a enum("a","a"));
create table t2 (a enum("ab","Ab"));
```

### What is changed and how it works?
Add check when ddl build column.

### Check List 

Tests 

 - Unit test

Code changes

 - Has interface methods change

Side effects

### CODE REVIEW

**Comment by crazycs520** [comment]
@winkyao @ciscoxll @zimulala PTAL

File: ddl/ddl_api.go [hunk 1]
```diff
@@ -551,6 +555,21 @@ func checkPriKeyConstraint(col *table.Column, hasDefaultValue, hasNullFlag bool,
 	return nil
 }
 
+func checkColumnValueConstraint(col *table.Column) error {
```
Comment (zimulala) [inline ddl/ddl_api.go]: s/checkColumnValueConstraint /checkEnumColumnValue ?

File: ddl/ddl_api.go [hunk 2]
```diff
@@ -551,6 +555,21 @@ func checkPriKeyConstraint(col *table.Column, hasDefaultValue, hasNullFlag bool,
...
+	if col.Tp != mysql.TypeEnum {
+		return nil
+	}
+	valueMap := make(map[string]struct{}, len(col.Elems))
+	for i := range col.Elems {
+		val := strings.ToLower(col.Elems[i])
+		if _, ok := valueMap[val]; ok {
+			return types.ErrDuplicatedValueInType.GenWithStackByArgs(col.Name, val, "ENUM")
```
Comment (zimulala) [inline ddl/ddl_api.go]: I think we need to replace `val` with `valueMap[val]`.
```
mysql> create table t1 (a enum("B","b"));
ERROR 1291 (HY000): Column 'a' has duplicated value 'B' in ENUM
```

File: ddl/ddl_api.go (hunk 1 above)
Comment (crazycs520) [reply ddl/ddl_api.go]: Now we should check set and enum.

Comment (zimulala) [reply ddl/ddl_api.go]: Before I comment it， it only check enum.

File: ddl/ddl_api.go [hunk 3]
```diff
@@ -551,6 +555,24 @@ func checkPriKeyConstraint(col *table.Column, hasDefaultValue, hasNullFlag bool,
 	return nil
 }
 
+func checkColumnValueConstraint(col *table.Column) error {
+	if col.Tp == mysql.TypeEnum || col.Tp == mysql.TypeSet {
+		valueMap := make(map[string]struct{}, len(col.Elems))
+		for i := range col.Elems {
+			val := strings.ToLower(col.Elems[i])
+			if _, ok := valueMap[val]; ok {
+				tpStr := "ENUM"
```
Comment (zimulala) [inline ddl/ddl_api.go]: I think it's better to put this code out of `for`.

File: ddl/ddl_api.go (hunk 2 above)
Comment (zimulala) [reply ddl/ddl_api.go]: No.
I mean this PR the error message like：
```
ERROR 1291 (HY000): Column 'a' has duplicated value 'b' in ENUM
```

File: ddl/ddl_api.go [hunk 4]
```diff
@@ -551,6 +555,24 @@ func checkPriKeyConstraint(col *table.Column, hasDefaultValue, hasNullFlag bool,
 	return nil
 }
 
+func checkColumnValueConstraint(col *table.Column) error {
+	if col.Tp == mysql.TypeEnum || col.Tp == mysql.TypeSet {
+		valueMap := make(map[string]string, len(col.Elems))
```
Comment (XuHuaiyu) [inline ddl/ddl_api.go]: use util.StringSet to replace `map[string]string`

File: ddl/db_test.go [hunk 5]
```diff
@@ -1528,6 +1528,23 @@ func (s *testDBSuite) TestCreateTable(c *C) {
...
+	s.testErrorCode(c, failSQL, tmysql.ErrDuplicatedValueInType)
+	// test for set column
+	failSQL = "create table t_enum (a set('e','e'));"
+	s.testErrorCode(c, failSQL, tmysql.ErrDuplicatedValueInType)
+	failSQL = "create table t_enum (a set('e','E'));"
+	s.testErrorCode(c, failSQL, tmysql.ErrDuplicatedValueInType)
+	failSQL = "create table t_enum (a set('abc','Abc'));"
+	s.testErrorCode(c, failSQL, tmysql.ErrDuplicatedValueInType)
+	_, err = s.tk.Exec("create table t_enum (a enum('B','b'));")
+	c.Assert(err.Error(), Equals, "[types:1291]Column 'a' has duplicated value 'B' in ENUM")
```
Comment (XuHuaiyu) [inline ddl/db_test.go]: why don't we use testErrorCode here?

File: ddl/ddl_api.go (hunk 4 above)
Comment (crazycs520) [reply ddl/ddl_api.go]: I need string value in map[string]string to generated error msg.

File: ddl/db_test.go (hunk 5 above)
Comment (crazycs520) [reply ddl/db_test.go]: I want to check check the error massage to make sure duplicated value is 'B' not 'b'.

File: ddl/ddl_api.go [hunk 6]
```diff
@@ -551,6 +555,24 @@ func checkPriKeyConstraint(col *table.Column, hasDefaultValue, hasNullFlag bool,
...
+		for i := range col.Elems {
+			val := strings.ToLower(col.Elems[i])
+			if _, ok := valueMap[val]; ok {
+				tpStr := "ENUM"
+				if col.Tp == mysql.TypeSet {
+					tpStr = "SET"
+				}
+				return types.ErrDuplicatedValueInType.GenWithStackByArgs(col.Name, valueMap[val], tpStr)
```
Comment (XuHuaiyu) [inline ddl/ddl_api.go]: why not use col.Elems[i] to replace valueMap[val] here?

Comment (crazycs520) [reply ddl/ddl_api.go]: In MySQL
```
mysql> create table t1 (a enum("B","b"));
ERROR 1291 (HY000): Column 'a' has duplicated value 'B' in ENUM
```
if  use col.Elems[i] to replace valueMap[val] here, the error msg return will be:
 ```
ERROR 1291 (HY000): Column 'a' has duplicated value 'b' in ENUM
```

File: ddl/ddl_api.go [hunk 7]
```diff
@@ -551,6 +555,24 @@ func checkPriKeyConstraint(col *table.Column, hasDefaultValue, hasNullFlag bool,
 	return nil
 }
 
+func checkColumnValueConstraint(col *table.Column) error {
+	if col.Tp == mysql.TypeEnum || col.Tp == mysql.TypeSet {
```
Comment (zz-jason) [inline ddl/ddl_api.go]: how about:
```go
if col.Tp == mysql.TypeEnum || col.Tp == mysql.TypeSet {
	return nil
}
// check the values.
```

**Comment by eurekaka** [comment]
`unit-test` is failing due to data race in explain analyze for hash join, looks weird.

**Comment by zz-jason** [comment]
Please cherry pick this PR to the release-2.1 and release-2.0 branch.

**Pull request merged**